### ISSUE TITLE
Generic dataclass inheritance false positive

### ISSUE THREAD

**Issue by tdsmith** [microsoft/pyright#9545]
An error is indicated on instantiation of a parameterized generic dataclass (Bookshelf) that subtypes a generic dataclass (Shelf) where the parent dataclass (Shelf) holds an instance of a generic type (Container) matching the parent dataclass’s parameterization:

```python
from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")

class Container(Generic[T]):
    pass

@dataclass
class Shelf(Generic[T]):
    container: Container[T] = Container[T]()

@dataclass
class Bookshelf(Shelf[T]):
    pass

# no error (as expected)
Shelf[int]()

# unexpected error in pyright 1.1.384 - 1.1.390
# Argument of type "Container[T@Shelf]" cannot be assigned to parameter "container" of type "Container[int]" in function "__init__"
#   "Container[T@Shelf]" is not assignable to "Container[int]"
#   Type parameter "T@Container" is invariant, but "T@Shelf" is not the same as "int"
Bookshelf[int]()
```

pyright 1.1.383 does not indicate any errors. pyright 1.1.384 flags errors on both the Shelf and Bookshelf instantiations. pyright 1.1.385 and 1.1.390 (latest) error on only the Bookshelf instantiation. Current mypy does not indicate an error.

No error is indicated for similar "plain" classes without `@dataclass` decorators where Shelf is provided with a simple constructor accepting a Container[T]:

```python
class PlainShelf(Generic[T]):
    container: Container[T]

    def __init__(self, container: Container[T] = Container[T]()):
        self.container = container

class PlainBookshelf(PlainShelf[T]):
    pass

# no error (as expected)
PlainShelf[int]()
# no error (as expected)
PlainBookshelf[int]()
```

I expected that T@Shelf would resolve to int when instantiating a Bookshelf[int].

Thanks for reviewing!

**Comment by tdsmith** [comment]
Thank you!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.391.
